### ISSUE TITLE
Missing files and functions are not accessible

### ISSUE THREAD

**Issue by Masterciw** [buildcenter/WindowsBuilder#2]
Hello,

first of all my respect for this crazy project.
When everything is working fine it would be a great tool
for customizing the windows 10 installation media.

Due my testing i found some i think minor problems due missing functions.
Here a short description of my setup

Fresh computer with Win10 1709 and latest updates.
Did change nothing on default configuration.

Followed instructions from walkthrough
Start Building with: build Win10-1709

First error in module LockScreen:
Anylyzed module LockScreen and found out that a function is missing:
Grant-AdminFullFileAccess

During applying Theme the following files are missing:
Resource file not found: aero_pin.cur
Resource file not found: aero_person.cur
Resource file not found: aero_pin_l.cur
Resource file not found: aero_person_l.cur

Next error on module OEMLicense
cmdlet or function: Grant-AdminFullFileAccess is missing

In module Package i also found a missing function called:
Grant-AdminFullRegistryKeyAccess

In module TTS a function called
Grant-AdminFullFileAccess is missing

In module Wallpaper once again function
Grant-AdminFullFileAccess is missing

Afterwards i got no other errors and build process finished like desired.

i think when the missing theme files are copied to the resource folder and
the missing functions are embedded everything works like desired

**Comment by imacks** [comment]
These functions:

```
Grant-AdminFullRegistryKeyAccess
Grant-AdminFullFileAccess
```

are in `tools\WindowsBuilder\WindowsBuildHelper.psm1`. I am not sure why it is failing to load in your case. Did you try `dir C:\WindowsBuilder -Recurse -File | Unblock-File`? Another thing I can think of is `Set-ExecutionPolicy Unrestricted` (you need to run that as admin). 

```
Resource file not found: aero_pin.cur
Resource file not found: aero_person.cur
Resource file not found: aero_pin_l.cur
Resource file not found: aero_person_l.cur
```

Guess I must have overlooked them. My `haro` theme cursors do not have pin and person, which only appeared in recent Windows versions. For my test, I just copied the ones from `C:\Windows\Cursors`. I will push some changes later to exclude these from the config file.

**Comment by Masterciw** [comment]
Okay, you,re right. The missing functions are inside the WindowsBuildHelper.psm1 but i can't imagine why they did not work.

Is there a way to view a logfile or an debugging mode which can be activated?
Even is there maybe a way to continue on error?

My execution-policy is set to unrestricted and i startet the script from an administrative cmd line.

During testing i found a typo in

Module BsodColor
Line 87 should be:
                        say ("Setting foreground color to {0}" -f $fgColor)
						
Maybe it also helps to list the possible colors in the bsd file.
So i tried to set it to blue but that is not available and throught an error.

Thats it for the moment. Have a nice day.

**Comment by Masterciw** [comment]
Did some more testing on a fresh installed computer with Windows 10 1709 and could reproduce the problem that modules with the functions

Grant-AdminFullRegistryKeyAccess
Grant-AdminFullFileAccess is missing

run on error.

Is there a way how can i help to troubleshoot this?

Greetings

**Comment by Masterciw** [comment]
Sorry i dont want to close this. My touchpad did not trigger the wanted action

**Comment by Masterciw** [comment]
... oh no i did it again. Sorry

**Comment by imacks** [comment]
Sorry I've been busy with other stuff. Can you try going in powershell and type: 

```powershell
ipmo c:\windowsbuilder\tools\WindowsBuilder\WindowsBuildHelper.psm1
get-command Grant-AdminFullRegistryKeyAccess
```

**Comment by Masterciw** [comment]
Thank you for your response. I'm also busy with some other tasks in the moment.
Tried this today and got the following response from PowerShell

CommandType Name                               Version  Source                                                                 
----------- ----                               -------  ------                                                                 
Function    Grant-AdminFullRegistryKeyAccess   0.0      WindowsBuildHelper 

The Import works well and the command could be found in the module.

What is the next step in troubleshooting this Issue?

**Comment by imacks** [comment]
In `src/Wallpaper/Install.ps1`, try adding this to the first line:
```
ipmo c:\windowsbuilder\tools\WindowsBuilder\WindowsBuildHelper.psm1
```

If you can confirm that the wallpaper module works with this, it must be a problem with the child scope not getting functions imported by the parent, and I can try fix it.

**Comment by Masterciw** [comment]
Added line:
ipmo c:\windowsbuilder\tools\WindowsBuilder\WindowsBuildHelper.psm1
before line
task default -depends Finalize
in module install.ps1 in folder wallpaper.

In bsd-file i only set the default wallpaper and did not removeextra


Then once again started the build process: Here is the output

Starting build module: Wallpaper
+----------+
| Precheck |
+----------+
+-----------+
| CopyFiles |
+-----------+
Replacing default
[0927:18] Error: In C:\WindowsBuilder\tools\Builder\Builder.psm1:357 Zeichen:25 +                         Die $_ 'InvokeTaskError' -NoWill +                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [<<==>>] Exception: Das Argument kann nicht an den Parameter "Message" gebunden werden, da es sich um eine leere Zeichenfolge handelt.

Translated it says something like: Exception: The argument could not be bound to the Parameter "Message" because the string is empty.

I'm trying to find out what [0927:18]  at the beginning of the line means. Do you now it and can maybe
decribe it to me.

Is perhaps the usage of a german Windows 10 Enterprise responsible for the errors.
I could setup another Computer with an english version.

So that are the 5 cents from my side right now. Thank you very much for your help and habe a nice evening.

**Comment by TotoTataJa** [comment]
Hi,
I'm testing this for the first time on the english US install.wim from Win10 1803 build.
I have the same message as Masterciw at the same line and position :

Starting build module: Package
Mounting registry hive to host: V:\WindowsBuilder-master\working\mount\Windows\System32\config\SOFTWARE --> HKLM:\svc_software
The operation completed successfully.

+----------+
| Precheck |
+----------+
[1155:11] Error: At V:\WindowsBuilder-master\tools\Builder\Builder.psm1:357 char:25 +                         Die $_ 'InvokeTaskError' -NoWill +                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [<<==>>] Exception: Cannot bind argument to parameter 'Message' because it is an empty string.

So Masterciw, ilt's not because of your German Windows 10 Ent...
Can you help us please ?

**Comment by imacks** [comment]
i've modified the upstream `builder` module; the `Cannot bind argument to parameter 'Message' ...` bug should be resolved by this. Hopefully this will help to debug things.

@TotoTataJa i got the same error on a new machine. Turns out it's a stupid problem with file attributes: make sure `install.wim` is not read-only, or mount fails :-)

**Comment by Masterciw** [comment]
Hello again,

did some more tests in the meantime

Installed 2 virtual machines. One with Windows 10 1709 in german and another one with Windows 10 1709 in english. Afterwards i donwloaded the latest master and configured the build environment.
As source install.wim i've got Windows 10 1709 in english and german.
Here are my test results:

1. On my english Windows 10 1709 test machine with english source everything is working well.

2. Tried on the same machine afterwards with the german sources and got the first error in the module OEMLicense and later in the module TTS. Both module are language depended.
In the module OEMLicense there is a hardcoded path which did not exist in the german wim file:
    $targetFile = Join-Path $BuildEnv.mountDir -ChildPath 'windows\System32\en-US\Licenses\Volume\Enterprise\license.rtf'
I check it in my environment. It is not enough to change en-US to de-DE because the folder layout is different in the german version
Same in the module TTS. Here are also lanaguage depending hardcoded paths are used in several places.

3. Next step is testing on a german Windows 1709 with english sources.
First error in module Lockscreen.
This module is using the function Grant-AdminFullFileAccess.
Deep dived into the Function Grant-AdminFullFileAccess.
Found the problem in the line:
	$ownerAccount = New-Object System.Security.Principal.NTAccount("BUILTIN\Administrators")
Like in the function Grant-AdminFullRegistryAccess there is a reference to the group "BUILTIN\Administrators" which did not exist in the german OS.
Next error in the module OEMLicense. Like in step 2 here is additional the problem with the function Grant-AdminFullFileAccess.

Next error is in the module PACKAGE in the function Grant-AdminFullRegistryAccess.
Error message is: Exception calling "AddAccessRule" with "1" argument(s): "Some or all identity references could not be translated"
When i look inside the function there is a reference to the group "BUILTIN\Administrators" This one did not exist in german windows.
There it is called "VORDEFINIERT\Administratoren".
I did a little research in the internet. In this case it should be better to use the SID because it is identically on each OS language.
Disabled the module and tested again. It then runs until the build process is finished without any other errors.
Next error in module TTS. Addtional to step 2 here is an error on function Grant-AdminFullFileAccess like i described before.
Next error in the module WALLPAPER. Error in fuction Grant-AdminFullFileAccess
Unfortunately my piece of paper with comments gets losts and due that the next errors are missing.

4. Last and least i tested it on a german Windows 1709 with german sources.
First error in module Lockscreen.
This module is using the function Grant-AdminFullFileAccess.
Deep dived into the Function Grant-AdminFullFileAccess.
Found the problem in the line:
	$ownerAccount = New-Object System.Security.Principal.NTAccount("BUILTIN\Administrators")
Like in the function Grant-AdminFullRegistryAccess there is a reference to the group "BUILTIN\Administrators" which did not exist in the german OS.
Next error in the module WALLPAPER. Error in fuction Grant-AdminFullFileAccess
Like in step 2 here is the next error in the module OEMLicense.
Disabled the module and did another run. Next error in module TTS.
When skipping this module the next error is in the module PACKAGE like in step 3.
Disabled the module for the next run. Due missing notices i can't remember if there where errors afterwards this step.

**Comment by imacks** [comment]
@Masterciw tks so much for this valuable feedback 😁. Now it is very safe to assume that the problem is culture related. I believe that replacing `BUILTIN\Administrators` with SIDs will make most of the problems go away. My bad!!!

Regarding the license file, would you be able to share with me the folder structure on your German system that leads to license.rtf?

**Comment by Masterciw** [comment]
**@imacks**  like desired here are the path for the license file on german windows 10 enterprise iso.
Checked this for 1703 und 1709 images.

1703 Windows\System32\de-DE\Licenses\Volume\Enterprise
1709 Windows\System32\de-DE\Licenses\Volume\Enterprise
Hope this will help you to fix that issue.

Drop me a line when additional testing is needed. Thx you once again for your great work on this project.

**Comment by imacks** [comment]
sorry this took a while. let me know if it fixed your issues 👍

**Comment by imacks** [comment]
I'm closing this for now. Feel free to open a new issue if the problem persists!
